crates/sources/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -1,57 +1,231 @@
+        for pool in registered_pools.pools {
+            let token0 = pool.token0.clone().context("token0 does not exist")?.id;
+            let token1 = pool.token1.clone().context("token1 does not exist")?.id;
+
+            let pair = TokenPair::new(token0, token1).context("cant create pair")?;
+            pools_by_token_pair.entry(pair).or_default().insert(pool.id);
+        }
+
         Ok(Self {
-            registered_pools: graph_api.get_registered_pools().await?,
+            pools_by_token_pair,
             graph_api,
+            cache: Mutex::new(Default::default()),
+            max_age,
         })
     }
+
+    async fn get_pools_and_update_cache(&self, pool_ids: &[H160]) -> Result<Vec<PoolData>> {
+        let pools = self.graph_api.get_pools_with_ticks_by_ids(pool_ids).await?;
+        let now = Instant::now();
+        let mut cache = self.cache.lock().unwrap();
+        for pool in &pools {
+            cache.insert(
+                pool.id,
+                CachedPool {
+                    pool: pool.clone(),
+                    updated_at: now,
+                    requested_at: now,
+                },
+            );
+        }
+        Ok(pools)
+    }
+
+    /// Returns cached pools and ids of outdated pools.
+    fn get_cached_pools(&self, token_pairs: &HashSet<TokenPair>) -> (Vec<PoolData>, Vec<H160>) {
+        let mut pool_ids = token_pairs
+            .iter()
+            .filter_map(|pair| self.pools_by_token_pair.get(pair))
+            .flatten()
+            .peekable();
+
+        match pool_ids.peek() {
+            Some(_) => {
+                let now = Instant::now();
+                let mut cache = self.cache.lock().unwrap();
+                pool_ids.partition_map(|pool_id| match cache.get_mut(pool_id) {
+                    Some(entry)
+                        if now.saturating_duration_since(entry.updated_at) < self.max_age =>
+                    {
+                        entry.requested_at = now;
+                        Either::Left(entry.pool.clone())
+                    }
+                    _ => Either::Right(pool_id),
+                })
+            }
+            None => Default::default(),
+        }
+    }
 }
 
 #[async_trait::async_trait]
 impl PoolFetching for UniswapV3PoolFetcher {
-    async fn fetch(&self, _token_pairs: HashSet<TokenPair>, _at_block: Block) -> Result<Vec<Pool>> {
-        todo!()
+    async fn fetch(&self, token_pairs: &HashSet<TokenPair>) -> Result<Vec<PoolData>> {
+        let (mut cached_pools, outdated_pools) = self.get_cached_pools(token_pairs);
+
+        if !outdated_pools.is_empty() {
+            let updated_pools = self.get_pools_and_update_cache(&outdated_pools).await?;
+            cached_pools.extend(updated_pools);
+        }
+
+        Ok(cached_pools)
+    }
+}
+
+pub struct CachingUniswapV3PoolFetcher(Arc<UniswapV3PoolFetcher>);
+
+impl CachingUniswapV3PoolFetcher {
```

Comment:
An alternative way to implement this could be to have a standalone top level function like

```rust
async fn update_pools_forever(
    inner: Weak<UniswapV3PoolFetcher>,
    update_interval: Duration,
    update_size: Option<usize>,
) 
```

which can then be spawned by the user of this module. This way you don't need a new struct, just one function. Not sure what is better, just wanted to mention it.

Reply:
Thanks for suggestion. Let's keep it as is since we might add additional maintenance task inside `AutoUpdatingUniswapV3PoolFetcher` for fetching the newly indexed pools as suggested in https://github.com/cowprotocol/cow-native-liquidity/pull/19#discussion_r899202137 and in that case having the struct would probably be cleaner.